pkgs/top-level/all-packages.nix

```diff
@@ -2755,7 +2755,21 @@ with pkgs;
 
   pcem = callPackage ../applications/emulators/pcem { };
 
-  pcsx2 = qt6Packages.callPackage ../applications/emulators/pcsx2 { };
+  pcsx2 = callPackage ../by-name/pc/pcsx2/package.nix {
+    shaderc = shaderc.overrideAttrs (old: {
+      patches = (old.patches or [ ]) ++ [
+        (fetchpatch {
+          url = "file://${pcsx2.src}/.github/workflows/scripts/common/shaderc-changes.patch";
+          hash = "sha256-Ps/D+CdSbjVWg3ZGOEcgbpQbCNkI5Nuizm4E5qiM9Wo=";
+          excludes = [
+            "CHANGES"
+            "CMakeLists.txt"
+            "libshaderc/CMakeLists.txt"
+          ];
+        })
+      ];
+    });
+  };
```

Comment:
IMO we should move that inside the package. Otherwise it is easily missed and forgotten.

Reply:
I agree with you but I don’t know ho to do this better? What is your solution to this?

When a custom dependency is needed, it should be passed just like this in `all-packages.nix`. This is how it is done for any derivation that needs different inputs. The derivation itself just uses whatever it transparently gets handed to. Of course here is more inline code due the patching, but there are also other packages doing it like this.

Patching an transparently passed dependency within the derivation (as initially proposed) is not a good solution (see my previous comment).

We could add a new non-existing input to the package with a default value (e.g. `shaderc-patched`). This way it is in pcsx2 package file and it could still be overridden by users. However if a new derivation is added named the same it will break. Further this breaks with the convention of how it is done with every other derivation.

So overall this seems to be the cleanest and best approach to me but I would be happy about suggestions.

Reply:
> When a custom dependency is needed, it should be passed just like this in `all-packages.nix`. This is how it is done for any derivation that needs different inputs. The derivation itself just uses whatever it transparently gets handed to. Of course here is more inline code due the patching, but there are also other packages doing it like this.

It can also be placed in a let in.

Reply:
Yes, but the problem is that the possibility of overriding `shaderc` is either lost or it does unexpected stuff with your own supplied `shaderc` package.

Initially (in this PR) the derivation got `shaderc` passed via an argument and patched it within the derivation. By doing so `shaderc` was theoretically overridable. It would however mess with the package supplied by the user and could fail in a patch conflict. Thus rendering this a bad solution because it does things where they should not be done leading to unexpected behavior.

In the duckstation PR `shaderc` is not passed as an argument (only indirectly via `callPackage`). So it is not overridable as usually, which is not great either.

The current solution makes the derivation behave as it should, without any unexpected behavior or caveats. Only problem is the patch that could easily be overlooked. So I suggest to fix this exact problem by simply putting a comment besides the `shaderc` argument that reminds about this fact.

Reply:
My rationale for duckstation was that the Duckstation developers are treating shaderc as a vendored dependency.
The difference is that they are not providing the whole Git subproject themselves, but a patch.

Then we should basically create a new package, `shaderc-specifically-patched-for-duckstation`.
But since this package is not meant to be useful outside duckstation, it is not exposed to top-level.